cmd/console-proxy/src/posix/picocom_map.rs

```diff
@@ -0,0 +1,185 @@
+                "crlf" => {
+                    ensure!(rules.cr.is_none(), "multiple rules remapping cr");
+                    rules.cr = Some(&[raw::LF]);
+                }
+                "crcrlf" => {
+                    ensure!(rules.cr.is_none(), "multiple rules remapping cr");
+                    rules.cr = Some(&[raw::CR, raw::LF]);
+                }
+                "igncr" => {
+                    ensure!(rules.cr.is_none(), "multiple rules remapping cr");
+                    rules.cr = Some(&[]);
+                }
+                "lfcr" => {
+                    ensure!(rules.lf.is_none(), "multiple rules remapping lf");
+                    rules.lf = Some(&[raw::CR]);
+                }
+                "lfcrlf" => {
+                    ensure!(rules.lf.is_none(), "multiple rules remapping lf");
+                    rules.lf = Some(&[raw::CR, raw::LF]);
+                }
+                "ignlf" => {
+                    ensure!(rules.lf.is_none(), "multiple rules remapping lf");
+                    rules.lf = Some(&[]);
+                }
+                "bsdel" => {
+                    rules.bsdel = true;
+                }
+                "delbs" => {
+                    rules.delbs = true;
+                }
+                _ => bail!("unknown or unsupported remap rule: {rule:?}"),
+            }
+        }
+
+        Ok(rules)
+    }
+}
+
+impl RemapRules {
+    pub fn apply<I>(&self, bytes: I) -> RemapIter<I::IntoIter>
+    where
+        I: IntoIterator<Item = u8>,
+    {
+        RemapIter {
+            inner: bytes.into_iter(),
+            prev: VecDeque::new(),
+            rules: *self,
+        }
+    }
+}
+
+pub struct RemapIter<I> {
+    inner: I,
+    prev: VecDeque<u8>,
+    rules: RemapRules,
+}
+
+impl<I> Iterator for RemapIter<I>
+where
+    I: Iterator<Item = u8>,
+{
+    type Item = u8;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            // Do we have any leftover bytes from a previous input byte
+            // (possibly from a previous iteration of this loop)?
+            if let Some(b) = self.prev.pop_front() {
+                return Some(b);
+            }
+
+            match self.inner.next()? {
+                raw::CR => {
+                    if let Some(repl) = self.rules.cr {
+                        self.prev.extend(repl);
+                        continue;
+                    } else {
+                        return Some(raw::CR);
+                    }
+                }
+                raw::LF => {
+                    if let Some(repl) = self.rules.lf {
+                        self.prev.extend(repl);
+                        continue;
+                    } else {
+                        return Some(raw::LF);
+                    }
+                }
+                raw::BS if self.rules.bsdel => return Some(raw::DEL),
+                raw::DEL if self.rules.delbs => return Some(raw::BS),
+                b => return Some(b),
+            }
```

Comment:
Nebulous vibes-based refactoring – it feels cleaner to _always_ push stuff to the ringbuf, even if we're going to pop it right away.  This lets us make the loop conditional on `is_empty` and move `pop_front` out of the loop entirely.
```suggestion
        // Do we have any leftover bytes from a previous input byte
        // (possibly from a previous iteration of this loop)?
        while self.prev.is_empty() {
            match self.inner.next()? {
                raw::CR if self.rules.cr.is_some() => {
                    self.prev.extend(self.rules.cr.unwrap());
                }
                raw::LF if self.rules.lf.is_some() => {
                    self.prev.extend(self.rules.lf.unwrap());
                }
                raw::BS if self.rules.bsdel => self.prev.push_back(raw::DEL),
                raw::DEL if self.rules.delbs => self.prev.push_back(raw::BS),
                b => self.prev.push_back(b),
            }
        }
        self.prev.pop_front()
```

What do you think – too clever?

Reply:
I don't think so; combined it with making `rules.{cr,lf}` non-optional to avoid the unwrap in 385d2d2